File: csharp/AdventOfCode.2021/Challenges/Challenge07.cs
using AdventOfCode.Core;
using AdventOfCode.Core.IO;
using AdventOfCode.Lib.Math;

namespace AdventOfCode2021.Challenges;

[Challenge(7)]
public class Challenge07
{
    private readonly IInputReader _inputReader;

    public Challenge07(IInputReader inputReader)
    {
        _inputReader = inputReader;
    }

    [Part1]
    public async Task<string> Part1Async()
    {
        var data = await _inputReader.ReadLineAsync<int>(7, ',').ToArrayAsync();

        var min = data.Min();
        var max = data.Max();

        var lowest = int.MaxValue;
        for (var pos = min; pos < max; pos++)
        {
            var test = data.Select(x => Math.Abs(x - pos)).Sum();
            if (test < lowest)
                lowest = test;
        }

        return lowest.ToString();
    }

    [Part2]
    public async Task<string> Part2Async()
    {
        var data = await _inputReader.ReadLineAsync<int>(7, ',').ToArrayAsync();

        var min = data.Min();
        var max = data.Max();

        var lowest = int.MaxValue;
        for (var pos = min; pos < max; pos++)
        {
            var test = data.Select(x => Euclid.TriangularNumber(Math.Abs(x - pos))).Sum();
            if (test < lowest)
                lowest = test;
        }

        return lowest.ToString();
    }
}

File: fsharp/AdventOfCode.2021/AdventOfCode.2021/Challenges/Challenge07.fs
module Challenges.Challenge07

open Utils.IO
open System

let triangularNumber n = (n*(n+1))/2

let setup = fun () -> readLine<int> 7 ','

let part1 input =
    let min = Array.min input
    let max = Array.max input

    [|min..max-1|]
    |> Seq.map (fun i -> input |> Seq.map (fun n -> Math.Abs (n-i)) |> Seq.sum)
    |> Seq.min
    |> string

let part2 input =
    let min = Seq.min input
    let max = Seq.max input

    [| min..max-1 |]
    |> Seq.map (fun i -> input |> Array.map (fun n -> (n-i) |> (Math.Abs >> triangularNumber)) |> Seq.sum)
    |> Seq.min
    |> string